src/Lib.P2P/SecureCommunication/Secio1Stream.cs

```diff
@@ -0,0 +1,270 @@
+            _inHmac.Init(new KeyParameter(localKey.MacKey));
+
+            _outHmac = new HMac(DigestUtilities.GetDigest(hashName));
+            _outHmac.Init(new KeyParameter(remoteKey.MacKey));
+
+            if (cipherName == "AES-256" || cipherName == "AES-512")
+            {
+                _decrypt = new CtrStreamCipher(new AesEngine());
+                var p = new ParametersWithIV(new KeyParameter(remoteKey.CipherKey), remoteKey.Iv);
+                _decrypt.Init(false, p);
+
+                _encrypt = new CtrStreamCipher(new AesEngine());
+                p = new ParametersWithIV(new KeyParameter(localKey.CipherKey), localKey.Iv);
+                _encrypt.Init(true, p);
+            }
+            else
+            {
+                throw new NotSupportedException($"Cipher '{cipherName}' is not supported.");
+            }
+        }
+
+        /// <inheritdoc />
+        public override bool CanRead => _stream.CanRead;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanWrite => _stream.CanRead;
+
+        /// <inheritdoc />
+        public override bool CanTimeout => false;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override void SetLength(long value) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+#pragma warning disable VSTHRD002 
+            return ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
+#pragma warning restore VSTHRD002 
+        }
+
+        /// <inheritdoc />
+        public override async Task<int> ReadAsync(byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+        {
+            var total = 0;
+            while (count > 0)
+
+                // Does the current packet have some unread data?
+                if (_inBlock != null && _inBlockOffset < _inBlock.Length)
+                {
+                    var n = Math.Min(_inBlock.Length - _inBlockOffset, count);
+                    Array.Copy(_inBlock, _inBlockOffset, buffer, offset, n);
+                    total += n;
+                    count -= n;
+                    offset += n;
+                    _inBlockOffset += n;
+                }
+
+                // Otherwise, wait for a new block of data.
+                else
+                {
+                    _inBlock = await ReadPacketAsync(cancellationToken);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)